libs/ml/include/ml/core/node.hpp

```diff
@@ -237,8 +237,10 @@ std::shared_ptr<TensorType> Node<TensorType>::Evaluate(bool is_training)
     {
       throw std::runtime_error("NaN encountered in Node::Evaluate");
     }
-
-    assert(!math::state_overflow<DataType>());
+    if (math::state_overflow<DataType>())
+    {
+      throw std::runtime_error("Overflow encountered in Node::Evaluate");
+    }
```

Comment:
I left the overflow checks as assertions because overflows are sometimes recoverable, unlike NaNs where the result of all operations will become NaN. Do you think it's better to have a runtime_error?

Reply:
the idea is: assert failure would anyway cause a runtime crash with unclear error message (because of messy core dump). In release build, no assertion would be made, but if overflows are not _always_ recoverable, I guess we need to stop VM here. However, it's just a guess, so if you are sure the assert is better here, I'll restore it.